docs/assigning-global-variables.md

```diff
@@ -0,0 +1,199 @@
+
+One example of such a decorator is:
+
+```typescript
+function registerEventHandler<TReturn, TArgs extends unknown[]>(
+  method: (...args: TArgs) => TReturn,
+  context: ClassMethodDecoratorContext,
+) {
+  /** @noSelf - the engine will not pass self parameter so wrap in lambda without self */
+  const contextless = (...args: TArgs) => method(...args);
+  // We can read the name of the method from the context
+  const globalName = context.name;
+  // @ts-ignore tell TS to ignore us 'illegally' writing to global scope
+  globalThis[globalName] = contextless;
+}
+```
+
+You can then write above example as:
+
+```typescript
+class EventHandlers {
+  @registerEventHandler
+  public OnStart() {
+    // start event handling code
+  }
+  @registerEventHandler
+  public OnStateChanged(newState: State) {
+    // state change event handler code
+  }
+}
+```
+
+:::note
+In the above example, `this` will be `nil` in the methods, do not try to use other members in the EventHandlers class!
+:::
+
+## Registering classes with a decorator
+
+Sometimes you want to register classes as globals, you can also do that with a decorator:
+
+```typescript
+function registerClass<TClass, TArgs extends unknown[]>(
+  c: new (...args: TArgs) => TClass,
+  context: ClassDecoratorContext,
+) {
+  if (context.name) {
+    // @ts-ignore tell TS to ignore us 'illegally' writing to global scope
+    globalThis[context.name] = c;
+  }
+}
+```
+
+You can now register any class by simply adding the decorator:
+
+```typescript
+@registerClass
+class EventHandlers {
+  public OnStart() {
+    // start event handling code
+  }
+  public OnStateChanged(newState: State) {
+    // state change event handler code
+  }
+}
+```
+
+### Custom global name decorator
+
+In the examples above, the decorators directly used the name of the decorated class or method, but with decorator parameters you can also specify custom override names:
+
+```typescript
+const registerClass =
+  (globalName: string) =>
+  <TClass, TArgs extends unknown[]>(c: new (...args: TArgs) => TClass, context: ClassDecoratorContext) => {
+    if (context.name) {
+      // @ts-ignore tell TS to ignore us 'illegally' writing to global scope
+      globalThis[globalName] = c;
+    }
+  };
+```
+
+Now instead of taking the global name from the class, you can specify a custom name yourself:
+
+```typescript
+@registerClass("CustomGlobalName")
+class EventHandlers {
+  public OnStart() {
+    // start event handling code
+  }
+  public OnStateChanged(newState: State) {
+    // state change event handler code
+  }
+}
+```
+
+## Assigning to globals with declarations
+
+The main weakness of the above methods is that you can declare any string, not protecting you from typos, and not giving any kind of editor support. This is fine if there are only a few such registrations that need to be done, but is somewhat error prone.
```

Comment:
For the first two options you also don't have to specify a string, right?

Also we should probably mention these issues earlier. Think it is important to highlight that editor support/type safety goes out the window as soon as you use any of the above solutions.